url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/util/io/UnicodeReader.java

```diff
@@ -0,0 +1,102 @@
+package fri.util.io;
+
+import java.io.*;
+
+/**
+ * Reads away UNICODE Byte Order Mark on construction. See
+ * http://www.unicode.org/unicode/faq/utf_bom.html
+ * 
+ * <pre>
+ *  00 00 FE FF    = UTF-32, big-endian
+ *  FF FE 00 00    = UTF-32, little-endian
+ *  FE FF          = UTF-16, big-endian
+ *  FF FE          = UTF-16, little-endian
+ *  EF BB BF       = UTF-8
+ * </pre>
+ */
+public class UnicodeReader extends Reader
+{
+	private static final int BOM_MAX_SIZE = 4;
+
+	private InputStreamReader delegate;
+
+	public UnicodeReader(InputStream in) throws IOException {
+		init(in, null);
+	}
+
+	public UnicodeReader(InputStream in, String defaultEnc) throws IOException {
+		init(in, defaultEnc);
+	}
+
+	/** Returns the encoding that was read from byte order mark if there was one. */
+	public String getEncoding() {
+		return delegate.getEncoding();
+	}
+
+	/**
+	 * Read-ahead four bytes and check for BOM marks. Extra bytes are unread back
+	 * to the stream, only BOM bytes are skipped.
+	 */
+	private void init(InputStream in, String defaultEnc) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 77-89 into a separate method.